File: go/src/leetcode/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
package leetcode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {

	ret := new(ListNode)
	head := ret
	total := 0
	carry := 0
	for l1 != nil || l2 != nil {
		if l1 != nil {
			total += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			total += l2.Val
			l2 = l2.Next
		}
		total += carry

		carry = total / 10

		head.Val = total % 10
		total = 0

		if l1 != nil || l2 != nil {
			head.Next = new(ListNode)
			head = head.Next
		}

	}
	if carry > 0 {
		head.Next = new(ListNode)
		head = head.Next
		head.Val = carry
	}
	return ret
}

File: archive/python/2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python
#
# [2] Add Two Numbers
#
# https://leetcode.com/problems/add-two-numbers/description/
#
# algorithms
# Medium (30.82%)
# Total Accepted:    810.9K
# Total Submissions: 2.6M
# Testcase Example:  '[2,4,3]\n[5,6,4]'
#
# You are given two non-empty linked lists representing two non-negative
# integers. The digits are stored in reverse order and each of their nodes
# contain a single digit. Add the two numbers and return it as a linked list.
#
# You may assume the two numbers do not contain any leading zero, except the
# number 0 itself.
#
# Example:
#
#
# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.
#
#
#
# Definition for singly-linked list.


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if not l1:
            return l2
        if not l2:
            return l1

        #  carry = total // 10
        carry = total = 0
        curr = head = l1
        l1_head = l1
        l2_head = l2

        while l1 or l2:
            total = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            carry = total // 10

            if l1:
                l1.val = total % 10
                l1 = l1.next
            if l2:
                l2.val = total % 10
                l2 = l2.next

            if l1 or l2:
                curr = l1 if l1 else l2
                head = l1_head if l1 else l2_head
            else:
                curr.next = None

        if carry > 0:
            curr.next = ListNode(carry)

        return head
